src/textual/widgets/_data_table.py

```diff
@@ -2108,12 +2108,17 @@ def _get_fixed_offset(self) -> Spacing:
     def sort(
         self,
         *columns: ColumnKey | str,
+        key: Callable | None = None,
```

Comment:
Since `key` can accept and return such a wide range of types were you thinking something like `key: Callable[..., Any]` or `key: Callable[[Any], Any]`? Neither are very "narrowing" but seem to be pretty common practice from what I have seen elsewhere.

Reply:
Doesn't the key function take a row, which is given as a list of values? In that case, wouldn't this be more precise?

```python
Callable[list[Any], Any]
```

I was hoping there was some kind of `typing.SupportsCompare`, but there doesn't seem to be.

Reply:
Yes, but the args going to `key` won't be a list as args come in as a tuple. And no matter the type, it seems the first type in the Callable typing has to be inside of brackets unless it is `...`. So, `Callable[[Any], Any]`, or `Callable[[list[Any], Any]`, or `Callable[[tuple[Any]], Any], and so on...

And to further complicate things, since itemgetter is used when columns are provided, either a single value `CellType` or tuple of multiple values `CellType` (when someone passes multiple columns to the `sort` function) are getting passed to the function making me lean towards `Any`.

Unless I am missing something (which is very possible), I think the options are `Callable[[Any], Any]` of `Callable[[tuple[Any]], Any]`. Thoughts?

Reply:
Please forgive my ignorance if this is out of line, would an `@overload` be appropriate to help narrow the type? Roughly mocked, it may look like the following:

```py
@overload
def sort(
    self,
    columns: ColumnKey | str,
    *,
    key: Callable[[CellType], Any] | None = None,
    reverse: bool = False,
) -> Self:
    ...

@overload
def sort(
    self,
    *columns: ColumnKey | str,
    key: Callable[[tuple[CellType, ...]], Any] | None = None,
    reverse: bool = False,
) -> Self:
    ...

def sort(
    self,
    *columns: ColumnKey | str,
    key: Callable[[Any], Any] | None = None,
    reverse: bool = False,
) -> Self:
    # Implementation
```

I do not fully understand the depth of `itemgetter()` so my suggestion may be erroneous.

Admittedly, there are two potential issues with my naive approach that I can see immediately:

1) I don't think the relocation of the `*` in the fist/singular `@overload` is good behavior. My type checker is yelling at me in the `_data_table.py` file on the implementation of `sort()` that `Overloaded implementation is not consistent with signature of overload 1` due to the movement of the `*` between the singular overload and the implementation of `sort()`. However, the type hints I'm receiving in `test_data_table.py` are appropriately enhanced and all of the updated/new `test_sort_...()` tests pass.
2) The first `@overload`, in its singular form, may become the first completion suggestion and how a new user first interacts with `sort()`. This may lead them to the incorrect conclusion that, despite `columns` being plural and clear documentation, `sort()` is limited to a single column input.

Reply:
Please forgive my ignorance if this is out of line, would an `@overload` be appropriate to help narrow the type? Roughly mocked, it could look like the following:

```py
@overload
def sort(
    self,
    columns: ColumnKey | str,
    *,
    key: Callable[[CellType], Any] | None = None,
    reverse: bool = False,
) -> Self:
    ...

@overload
def sort(
    self,
    *columns: ColumnKey | str,
    key: Callable[[tuple[CellType, ...]], Any] | None = None,
    reverse: bool = False,
) -> Self:
    ...

def sort(
    self,
    *columns: ColumnKey | str,
    key: Callable[[Any], Any] | None = None,
    reverse: bool = False,
) -> Self:
    # Implementation
```

Though the type hints I'm receiving in `test_data_table.py` are appropriately enhanced and the new/updated `test_sort_...()` tests are still passing on my machine, there are two potential issues with my naive approach that I can see immediately:

1) These overloads are just fancy lies and I don't think the relocation of the `*` in the fist/singular `@overload` is good behavior. My type checker is yelling at me in the `_data_table.py` file on the implementation of `sort()` that `Overloaded implementation is not consistent with signature of overload 1` due to the movement of the `*` between the singular overload and the implementation. This motion hides the fact that the singular `columns: ColumnKey | str` param is actually getting wrapped up into a tuple by the implementation's `*columns: ColumnKey | str`.
2) The first `@overload`, in its singular form, may become the first completion suggestion and how a new user first interacts with `sort()`. This may lead them to the incorrect conclusion that, despite `columns` being plural and clear documentation, `sort()` is limited to a single column input.

Reply:
I honestly haven't used the `@overload` decorator much and that would probably be a directional decision for the maintainers.

As for the `Callable` typing... I have looked into it (too much) and there doesn't seem to be a general consensus. After lots of GitHub sleuthing, I have found most packages just go with `Callable[[Any], Any]` or `Callable[..., Any]` which [is equivalent to just](https://docs.python.org/3.9/library/typing.html#typing.Callable) `Callable`. And I can't come up with anything better...

On the inputs/parameters side of the callable type, `CellType` is loosely defined so no matter if you use that, it just resolves to `Any` anyway.

And on the return side of the callable, since the method is called `sort`, I think it is implied that the return values should be "comparable", but that obviously leaves many options (which is probably best typed with `Any`).

So, all that to say, 🤷‍♂️...